src/algorithms/nolip.jl

```diff
@@ -0,0 +1,223 @@
+    z_curr::Tx = similar(x)
+    Az::TAx = similar(Ax)
+    grad_f_Az::TAx = similar(Ax)
+    At_grad_f_Az::Tx = similar(x)
+end
+
+f_model(iter::NOLIPIteration, state::NOLIPState) = f_model(state.f_Ax, state.At_grad_f_Ax, state.res, iter.alpha / state.gamma)
+
+function Base.iterate(iter::NOLIPIteration{R}) where {R}
+    x = copy(iter.x0)
+    Ax = iter.A * x
+    grad_f_Ax, f_Ax = gradient(iter.f, Ax)
+    gamma = iter.gamma === nothing ? iter.alpha / lower_bound_smoothness_constant(iter.f, iter.A, x, grad_f_Ax) : iter.gamma
+    At_grad_f_Ax = iter.A' * grad_f_Ax
+    y = x - gamma .* At_grad_f_Ax
+    z, g_z = prox(iter.g, y, gamma)
+    state = NOLIPState(
+        x=x, Ax=Ax, f_Ax=f_Ax, grad_f_Ax=grad_f_Ax, At_grad_f_Ax=At_grad_f_Ax,
+        gamma=gamma, y=y, z=z, g_z=g_z, res=x-z, H=initialize(iter.directions, x),
+    )
+    if (iter.gamma === nothing || iter.adaptive == true)
+        state.gamma, state.g_z, f_Az, f_Az_upp = backtrack_stepsize!(
+            state.gamma, iter.f, iter.A, iter.g,
+            state.x, state.f_Ax, state.At_grad_f_Ax, state.y, state.z, state.g_z, state.res,
+            state.Az, state.grad_f_Az,
+            alpha = iter.alpha, minimum_gamma = iter.minimum_gamma,
+        )
+    end
+    return state, state
+end
+
+set_next_direction!(::QuasiNewtonStyle, ::NOLIPIteration, state::NOLIPState) = mul!(state.d, state.H, -state.res)
+set_next_direction!(::NoAccelerationStyle, ::NOLIPIteration, state::NOLIPState) = state.d .= .-state.res
+set_next_direction!(iter::NOLIPIteration, state::NOLIPState) = set_next_direction!(acceleration_style(typeof(iter.directions)), iter, state)
+
+update_direction_state!(::QuasiNewtonStyle, ::NOLIPIteration, state::NOLIPState) = update!(state.H, state.x - state.x_prev, state.res - state.res_prev)
+update_direction_state!(::NoAccelerationStyle, ::NOLIPIteration, state::NOLIPState) = return
+update_direction_state!(iter::NOLIPIteration, state::NOLIPState) = update_direction_state!(acceleration_style(typeof(iter.directions)), iter, state)
+
+reset_direction_state!(::QuasiNewtonStyle, ::NOLIPIteration, state::NOLIPState) = reset!(state.H)
+reset_direction_state!(::NoAccelerationStyle, ::NOLIPIteration, state::NOLIPState) = return
+reset_direction_state!(iter::NOLIPIteration, state::NOLIPState) = reset_direction_state!(acceleration_style(typeof(iter.directions)), iter, state)
+
+function Base.iterate(iter::NOLIPIteration{R}, state::NOLIPState) where R
+    f_Az, a, b, c = R(Inf), R(Inf), R(Inf), R(Inf)
+
+    # store iterate and residual for metric update later on
+    state.x_prev .= state.x
+    state.res_prev .= state.res
+
+    # compute FBE
+    FBE_x = f_model(iter, state) + state.g_z
+
+    sigma = iter.beta * (0.5 / state.gamma) * (1 - iter.alpha)
+    tol = 10 * eps(R) * (1 + abs(FBE_x))
+    threshold = FBE_x - sigma * norm(state.res)^2 + tol
+
+    tau_backtracks = 0
+    can_update_direction = true
+
+    while true
+
+        if can_update_direction
```

Comment:
Minor: this is more of a `should_` than a `can_`?

Reply:
I think this gives more emphasis on the fact that `d` cannot be updated sometimes. Conversely, as argued in the preprint, the search direction is updated whenever it `can_` be, but this is not necessary.